### PR TITLE
Started on V2, not using ASTs...

### PR DESCRIPTION
**Pull request by matanlurey**
... as requested by the Dart analyzer/front end team.

Started using built_value (since it gives it a nicer API anyway) and modeled method bodies as code blocks instead, since in reality it makes writing code generation that much easier if you only use typed structures for the API.

[I realize this change is scary, hence why it's going on the `V2` branch instead until complete]

### CODE REVIEW

File: lib/src/emitter.dart [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    visitCode(spec.code, output);
+    output.write(' ');
+    return output;
+  }
+
+  @override
+  visitClass(Class spec, [StringSink output]) {
+    output ??= new StringBuffer();
+    spec.docs.forEach(output.writeln);
+    spec.annotations.forEach((a) => visitAnnotation(a, output));
```
Comment (alorenzen) [inline lib/src/emitter.dart]: Avoid using forEach - https://www.dartlang.org/guides/language/effective-dart/usage#avoid-using-iterableforeach-with-a-function-literal

File: lib/src/emitter.dart [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        ..writeAll(spec.mixins.map<StringSink>(visitType), ',');
+    }
+    if (spec.implements.isNotEmpty) {
+      output
+        ..write(' implements ')
+        ..writeAll(spec.implements.map<StringSink>(visitType), ',');
+    }
+    output.write(' {');
+    spec.constructors.forEach((c) => visitConstructor(c, spec.name, output));
+    spec.fields.forEach((f) => visitField(f, output));
```
Comment (alorenzen) [inline lib/src/emitter.dart]: Same here, just use a for-loop.

File: lib/src/visitors.dart [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  T visitSpec(Spec spec) => spec.accept(this);
+
+  @override
+  T visitType(TypeReference spec) => null;
+
+  @override
+  T visitTypeParameters(Iterable<TypeReference> specs) => null;
+}
+
+class RecursiveSpecVisitor<T> extends SimpleSpecVisitor<T> {
```
Comment (alorenzen) [inline lib/src/visitors.dart]: TODO here for implementation?

**Comment by matanlurey** [comment]
@alorenzen Yeah, I think I can re-add scoping and imports in the next PR, just wanted to make sure I was covering all feature requirements for the injection package first so I can land this internally.

File: lib/src/emitter.dart (hunk 2 above)
Comment (matanlurey) [inline lib/src/emitter.dart]: Will fix style nits in next PR.

File: lib/src/emitter.dart (hunk 1 above)
Comment (matanlurey) [inline lib/src/emitter.dart]: Will fix style nits in next PR.

File: lib/src/visitors.dart (hunk 3 above)
Comment (matanlurey) [inline lib/src/visitors.dart]: Will fix or remove in next PR.

**Pull request merged**